src/systems/environment_preload/VisualizationTool.hh

```diff
@@ -0,0 +1,141 @@
+{
+inline namespace GZ_SIM_VERSION_NAMESPACE
+{
+
+/// \brief This class helps handle point cloud visuallizations
+/// of environment data.
+class EnvironmentVisualizationTool
+{
+  /// \brief Environment constructor
+  public: EnvironmentVisualizationTool();
+
+  /// \brief To synchronize member access.
+  private: std::mutex mutex;
+
+  /// \brief First load we need to scan for existing data sensor
+  private: bool first {true};
+
+  /// \brief Enable resampling
+  public: std::atomic<bool> resample{true};
+
+  /// \brief Time has come to an end.
+  private: bool finishedTime{false};
+
+  /// \brief Create publisher structures whenever a new environment is made
+  /// available.
+  /// \param[in] _data Data to be visuallized
+  private: void CreatePointCloudTopics(
+    const std::shared_ptr<components::EnvironmentalData> _data);
+
+  /// \brief Invoke when new file is made available.
+  public: void FileReloaded();
+
+  /// \brief Step the visualizations
+  /// \param[in] _info The simulation info including timestep
+  /// \param[in] _ecm The Entity-Component-Manager
+  /// \param[in] _data The data to be visuallized
+  /// \param[in] _xSample Samples along x
+  /// \param[in] _ySample Samples along y
+  /// \param[in] _zSample Samples along z
+  public: void Step(
+    const UpdateInfo &_info,
+    const EntityComponentManager& _ecm,
+    const std::shared_ptr<components::EnvironmentalData> _data,
+    double _xSamples, double _ySamples, double _zSamples);
+
+  /// \brief Publishes a sample of the data
+  /// \param[in] _data The data to be visuallized
+  /// \param[in] _xSample Samples along x
+  /// \param[in] _ySample Samples along y
+  /// \param[in] _zSample Samples along z
+  private: void Visualize(
+    const std::shared_ptr<components::EnvironmentalData> data,
+    double _xSamples, double _ySamples, double _zSamples);
+
+  /// \brief Get the point cloud data.
+  private: void Publish();
+
+  /// \brief Resize the point cloud structure (used to reallocate
+  /// memory when resolution changes)
+  /// \param[in] _ecm The Entity-Component-Manager
+  /// \param[in] _data The data to be visuallized
+  /// \param[in] _xSample Samples along x
+  /// \param[in] _ySample Samples along y
+  /// \param[in] _zSample Samples along z
+  private: void ResizeCloud(
+    const std::shared_ptr<components::EnvironmentalData> _data,
+    const EntityComponentManager& _ecm,
+    unsigned int _xSamples, unsigned int _ySamples, unsigned int _zSamples);
+
+  /// \brief Publisher for point clouds
+  private: transport::Node::Publisher pcPub;
+
+  /// \brief Publishers for data
+  private: std::unordered_map<std::string, transport::Node::Publisher> pubs;
+
+  /// \brief Floating point message buffers
+  private: std::unordered_map<std::string, gz::msgs::Float_V> floatFields;
+
+  /// \brief GZ buffers
+  private: transport::Node node;
+
+  /// \brief Point cloud buffer
+  private: gz::msgs::PointCloudPacked pcMsg;
+
+  /// \brief Session cursors
```

Comment:
What's a session, the key is a timestamp?

Reply:
We use sessions to keep track of the current time (the reason is simply efficiency): https://github.com/gazebosim/gz-math/blob/836a5c504255f3b309bdbcdb24fdf25bc9b9423c/include/gz/math/TimeVaryingVolumetricGridLookupField.hh#L48